src/Storages/StorageMergeTree.cpp

```diff
@@ -1240,56 +1240,63 @@ MergeTreeDataPartPtr StorageMergeTree::outdatePart(const String & part_name, boo
     }
 }
 
-void StorageMergeTree::dropPartition(const ASTPtr & partition, bool detach, bool drop_part, ContextPtr local_context, bool throw_if_noop)
+void StorageMergeTree::dropPartNoWaitNoThrow(const String & part_name)
 {
-    {
-        MergeTreeData::DataPartsVector parts_to_remove;
-        auto metadata_snapshot = getInMemoryMetadataPtr();
+    if (auto part = outdatePart(part_name, false))
+        dropPartsImpl({part}, false);
 
-        if (drop_part)
-        {
-            auto part = outdatePart(partition->as<ASTLiteral &>().value.safeGet<String>(), throw_if_noop);
-            /// Nothing to do, part was removed in some different way
-            if (!part)
-                return;
+    /// Else nothing to do, part was removed in some different way
+}
 
-            parts_to_remove.push_back(part);
-        }
-        else
-        {
-            /// Asks to complete merges and does not allow them to start.
-            /// This protects against "revival" of data for a removed partition after completion of merge.
-            auto merge_blocker = stopMergesAndWait();
-            String partition_id = getPartitionIDFromQuery(partition, local_context);
-            parts_to_remove = getDataPartsVectorInPartition(MergeTreeDataPartState::Committed, partition_id);
-
-            /// TODO should we throw an exception if parts_to_remove is empty?
-            removePartsFromWorkingSet(parts_to_remove, true);
-        }
+void StorageMergeTree::dropPart(const String & part_name, bool detach, ContextPtr /*query_context*/)
+{
+    if (auto part = outdatePart(part_name, true))
+        dropPartsImpl({part}, detach);
 
-        if (detach)
-        {
-            /// If DETACH clone parts to detached/ directory
-            /// NOTE: no race with background cleanup until we hold pointers to parts
-            for (const auto & part : parts_to_remove)
-            {
-                LOG_INFO(log, "Detaching {}", part->relative_path);
-                part->makeCloneInDetached("", metadata_snapshot);
-            }
-        }
+    /// Else nothing to do, part was removed in some different way
+}
 
-        if (deduplication_log)
+void StorageMergeTree::dropPartition(const ASTPtr & partition, bool detach, ContextPtr local_context)
+{
+    /// Asks to complete merges and does not allow them to start.
+    /// This protects against "revival" of data for a removed partition after completion of merge.
+    auto merge_blocker = stopMergesAndWait();
```

Comment:
it's a pity that we can't do it per partition... Stopping (and waiting) of all merges in 'hot' partition during dropping some old partition is not great. And while files are removed no merges will start. 

Maybe we can release the merge_blocker bit earlier - for ex. after removePartsFromWorkingSet ?

Reply:
Yes, lock can released after `removePartsFromWorkingSet`. It's better to lock for merges only parts, that are currently removing, but not in this PR.